### DISCUSSION TITLE
Debug mode

### DISCUSSION THREAD

**Discussion by rwadowski** [lf-edge/ekuiper#908] [Q&A] [answered]
I am starting to work on this project and I have question - how to debug the codebase ? - I want to use JetBrains Goland and ideally I want to run the app with the ide in debug mode. I did not find too much info how to run it locally with minimal run. I am not golang expert (I com from the jvm world) so maybe it is not the golang way.

**Comment by superrxan** [comment]
Hi rwadowski

There is a quick method to play around with eKuiper locally. please refer following steps : 
1. Build the binary executable `kuiper` and `kuiperd`, where `kuiperd` is the process engine and `kuiper` is cli client. 
    you can download the source code and just run `make`, these files will be built and put into the `_build/kuiper-*/bin/`
2. Set eKuiper source to an MQTT server. This sample uses server locating at ``tcp://broker.emqx.io:1883``. ``broker.emqx.io`` is a public MQTT test server hosted by [EMQ](https://www.emqx.io); also set `kuiperd` to debug mode . Then run the `kuiperd`
    ```shell
     export MQTT_SOURCE__DEFAULT__SERVERS="[tcp://broker.emqx.io:1883]"
     export KUIPER__BASIC__DEBUG="true"
     #go to the _build/kuiper-*/ and run kuiperd
     ./bin/kuiperd
    ```
3. Create a stream - the stream is your stream data schema, similar to table definition in database. Let's say the temperature & humidity data are sent to ``broker.emqx.io``, and those data will be processed in your **LOCAL RUN** eKuiper docker instance.  Below steps will create a stream named ``demo``, and data are sent to ``devices/device_001/messages`` topic, while ``device_001`` could be other devices, such as ``device_002``, all of those data will be subscribed and handled by ``demo`` stream.

   ```shell
   # open another shell ,go to the _build/kuiper-*/ and run kuiper cli
   ./bin/kuiper create stream demo '(temperature float, humidity bigint) WITH (FORMAT="JSON",DATASOURCE="devices/+/messages")'
   Connecting to 127.0.0.1:20498...
   Stream demo is created.

   ```
4. Create a rule by sql, in the same terminal with step3 and type    
    ```
     ./bin/kuiper query
     kuiper > select * from demo where temperature > 30;
      Query was submit successfully.
     ```
5.  Publish sensor data to topic ``devices/device_001/messages`` of server ``tcp://broker.emqx.io:1883`` with any [MQTT client tools](https://medium.com/@emqtt/mqtt-client-tools-215ff7a17ad). You can also use [MQTTX](https://mqttx.app) . Below sample uses ``mosquitto_pub``. 
     ```shell
     mosquitto_pub -h broker.emqx.io -m '{"temperature": 40, "humidity" : 20}' -t devices/device_001/messages
   
     ```   
    
6.  If everything goes well,  you can see the message is print on docker ``bin/kuiper query`` window. Please try to publish another message with ``temperature`` less than 30, and it will be filtered by WHERE condition of the SQL. 

     ```
     kuiper > select * from demo WHERE temperature > 30;
     [{"temperature": 40, "humidity" : 20}]
     ```
7.  you can check the logs in the log directory to debug 

Reference:
     https://github.com/lf-edge/ekuiper/blob/master/deploy/docker/README.md

**Comment by ngjaying** [comment, marked as answer]
Thanks @SuperRxan . I will add some tips about debugging in Goland.  The general usage instruction can be found on [GoLand help](https://www.jetbrains.com/help/go/debugging-code.html#general-procedure). Specific to eKuiper, you can debug with the same procedure. Just need to find out the entry program:

1. Debug the whole program. Make sure `data`, `plugins/sources`, `plugins/functions`, `plugins/sinks`, , `logs` are created in the eKuiper root path. Add your breakpoints. Open `cmd/kuiperd/main.go`. In the main function, you'll find a green triangle in the ruler, click it and select debug. Then create your stream/rule that would run through your breakpoint, the debugger will pause there.
2. To debug a small portion of code, I recommend to write a unit test and debug it. You can go to any test file and find the same green triangle to run in debug mode. For example, `pkg/cast/cast_test.go` `TestMapConvert_Funcs` can run as debug.

**Reply by rwadowski** [reply]
Hi, thanks for answer. I have question about data, plugins and log directory. I think that the root of eKuiper is set by KuiperBaseKey env variable. Is there default value I should pass to the kuiperd ? When I try to pass ~/opensource/repositories/ekuiper/ (place where I have project) it crashes because path resolved for example for logs is /log instead of ~/opensource/repositories/ekuiper/logs - my KuiperBaseKey variable is set to absolute path of ~/opensource/repositories/ekuiper/. Is it correct setting ?

**Reply by ngjaying** [reply]
`KuiperBaseKey` is only adopted in abstract path mode which is set in CI when built into rpm or deb. The default mode is relative path, so you'll just need to create those directories in the root path.

**Reply by rwadowski** [reply]
So the root path is the path of the project ?

**Reply by ngjaying** [reply]
yes

**Reply by rwadowski** [reply]
I have deleted KuiperBaseKey env variable and I have something like this `conf dir not found, please set KuiperBaseKey program environment variable correctly.`

**Reply by ngjaying** [reply]
make sure you have all the directories listed in line 19 of the [Makefile](https://github.com/lf-edge/ekuiper/blob/master/Makefile)

**Reply by rwadowski** [reply]
awesome it worked and I was able to catch breakpoint, thanks for help
